src/paillier.rs

```diff
@@ -61,15 +63,58 @@ impl PaillierDecryptionKey {
         Ok(self.0.decrypt(c).ok_or(PaillierError::DecryptionFailed)?)
     }
 
-    /// Generate a new [`PaillierDecryptionKey`] from two primes.
+    /// Generate a new [`PaillierDecryptionKey`] and its factors.
     ///
-    /// The inputs `p` and `q` are checked for primality but not for _safe_ primality.
-    /// TODO #56: Check prime size and output size.
-    pub(crate) fn from_primes(p: &BigNumber, q: &BigNumber) -> Result<Self> {
-        Ok(PaillierDecryptionKey(
-            libpaillier::DecryptionKey::with_primes(p, q)
+    /// The factors `p` and `q` are `PRIME_BITS`-long safe primes, and the resulting
+    /// modulus is `2 * PRIME_BITS` long.
+    pub(crate) fn new(
+        rng: &mut (impl CryptoRng + RngCore),
+    ) -> Result<(Self, BigNumber, BigNumber)> {
+        // Local utility function to generate prime pairs and check their resulting modulus length.
+        let generate_prime_pair = || {
```

Comment:
I'm a fan of local functions! I think this one is big enough to be a named local function. This way we can give it a signature:
```diff
- let generate_prime_pair = || {
+ fn generate_prime_pair() -> Option<(BigNumber, BigNumber)> {
```

Reply:
I like that idea but the signature is kind of ugly because this actually needs to take an `rng`. Does this change your preference? I ended up with
```rust
    fn generate_prime_pair(
        rng: &mut (impl CryptoRng + RngCore),
    ) -> impl FnMut() -> Option<(BigNumber, BigNumber)> + '_ { ... }
```

Reply:
Oh yeah, that's way uglier :mask: . We should just keep the original code! I forgot about the `rng` being captured by the closure. NVM!